### PR TITLE
0.13.0 PLAT 5120

### PR DESCRIPTION
**Pull request by lilachmaliniak**

### CODE REVIEW

**Comment by kaltura-hooks** [comment]
Hi @lilachmaliniak,
Thank you for contributing this pull request!
Please sign the Kaltura CLA so we can review and merge your contribution.
Learn more at http://bit.ly/KalturaContrib

File: dvr-ws/routes/index.js [hunk 1]
```diff
@@ -2,18 +2,53 @@ var express = require('express');
...
+                res.status(404).send('File expired');
+        }
+    }
+    else {
+        res.sendFile(fullPath);
+    }
 
 });
 
+function checkIfFileExpired(fullPath, fileExpireAge) {
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: checkExpriedAndSendFile

File: lib/manifest/EndListAppender.js [hunk 2]
```diff
@@ -24,31 +25,53 @@ var EndlistAppender = function EndListAppender(){
...
-                        }
-                        manifest.set('playlistType', 'VOD');
-                        return qio.write(flavorManifestPath, manifest.toString());
-                    })
-                    .catch(function (err) {
-                        logger.error("[%s] Adding EXT-X-ENDLIST at location: %s failed; Error: %s", entryId, flavorManifestPath, ErrorUtils.error2string(err));
-                    });
+        var deferred = Q.defer();
+
+        backendClient.getLiveEntryServerNodes(entryId)
```
Comment (kalturaguy) [inline lib/manifest/EndListAppender.js]: return backendClient.getLiveEntryServerNodes

File: lib/manifest/EndListAppender.js [hunk 3]
```diff
@@ -24,31 +25,53 @@ var EndlistAppender = function EndListAppender(){
...
+            .then( function(liveEntryServerNodes) {
+                if (liveEntryServerNodes.count > 0) {
+                    logger.debug("[%s], the live stream has at least one providing data center", entryId);
+                    return deferred.reject();
```
Comment (kalturaguy) [inline lib/manifest/EndListAppender.js]: return Q.reject()

File: lib/manifest/EndListAppender.js [hunk 4]
```diff
@@ -58,7 +81,7 @@ var EndlistAppender = function EndListAppender(){
                 logger.info("[%s] Timer is already pending, ignoring request", entryId);
                 return;
             }
-            var timeoutObj = setTimeout(function () {
+            var timeoutObj = setInterval(function () {
                 logger.debug("[%s] Timer ended. Starting endlistAppender", entryId);
```
Comment (kalturaguy) [inline lib/manifest/EndListAppender.js]: avoid re-entrance

File: lib/manifest/EndListAppender.js [hunk 5]
```diff
@@ -58,7 +81,7 @@ var EndlistAppender = function EndListAppender(){
                 logger.info("[%s] Timer is already pending, ignoring request", entryId);
                 return;
             }
-            var timeoutObj = setTimeout(function () {
+            var timeoutObj = setInterval(function () {
```
Comment (kalturaguy) [inline lib/manifest/EndListAppender.js]: change to settimeout with catch processing to redo the settimeout

File: dvr-ws/routes/index.js (hunk 1 above)
Comment (lilachmaliniak) [inline dvr-ws/routes/index.js]: done.

File: lib/manifest/EndListAppender.js (hunk 2 above)
Comment (lilachmaliniak) [inline lib/manifest/EndListAppender.js]: done.

File: lib/manifest/EndListAppender.js (hunk 3 above)
Comment (lilachmaliniak) [inline lib/manifest/EndListAppender.js]: done.

File: dvr-ws/routes/index.js [hunk 6]
```diff
@@ -2,18 +2,48 @@ var express = require('express');
...
+                 logger.debug('modified time of %s is %s. File is valid.', fullPath, stats.mtime.toDateString());
+                 res.sendFile(fullPath);
+             }
+             else {
+                 logger.warn('modified time of %s is %s. File expired.', fullPath, stats.mtime.toDateString());
+                 res.status(404).send('File expired');
+             }
+         });
+    } catch (e) {
+        logger.error('exception, failed to get modified time of %s. error: %s', fullPath, errorUtils.error2string(e));
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: should call     res.status(404).send('XXXX');

File: dvr-ws/routes/index.js [hunk 7]
```diff
@@ -2,18 +2,48 @@ var express = require('express');
...
 
+function checkExpriedAndSendFile(fullPath, res) {
+
+    var oldestValidAge = Date.now() - chunklistExpireAge;
+
+    try {
+         fs.stat(fullPath, function(err, stats) {
+
+             if (stats.mtime.getTime() >= oldestValidAge) {
+                 logger.debug('modified time of %s is %s. File is valid.', fullPath, stats.mtime.toDateString());
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: please comment this

File: dvr-ws/routes/index.js [hunk 8]
```diff
@@ -2,18 +2,48 @@ var express = require('express');
...
 });
 
+function checkExpriedAndSendFile(fullPath, res) {
+
+    var oldestValidAge = Date.now() - chunklistExpireAge;
+
+    try {
+         fs.stat(fullPath, function(err, stats) {
+
+             if (stats.mtime.getTime() >= oldestValidAge) {
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: please check if err

File: lib/manifest/EndListAppender.js [hunk 9]
```diff
@@ -24,51 +23,74 @@ var EndlistAppender = function EndListAppender(){
...
+        return backendClient.getLiveEntryServerNodes(entryId)
+            .then( function(liveEntryServerNodes) {
+                if (liveEntryServerNodes.count > 0) {
+                    var err = String.format("[%s], the live stream has at least one providing data center", entryId);
+                    return Q.reject(err);
+                }
+            })
+            .catch( function () {
+                var err = String.format("[%s] failed to retrieve server nodes", entryId);
+                return Q.reject(err);
```
Comment (kalturaguy) [inline lib/manifest/EndListAppender.js]: err?

**Pull request merged**